crates/optimism/primitives/src/op_tx_type.rs

```diff
@@ -0,0 +1,198 @@
+//! newtype pattern on op_alloy_consensus::OpTxType.
+//! OpTxType that implements reth_primitives_traits::TxType.
+//! This type is required because a Compact impl is needed on the deposit tx type.
+
+use alloy_eips::eip2718::Eip2718Error;
+use alloy_primitives::{U64, U8};
+use alloy_rlp::{Decodable, Encodable};
+use bytes::BufMut;
+use core::fmt::Debug;
+use derive_more::{
+    derive::{From, Into},
+    Display,
+};
+use op_alloy_consensus::OpTxType as AlloyOpTxType;
+use reth_primitives_traits::TxType;
+use std::convert::TryFrom;
+
+/// Wrapper type for `AlloyOpTxType` to implement `TxType` trait.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Display, Ord, Hash, From, Into)]
+#[from(AlloyOpTxType)]
+#[into(u8)]
+pub struct OpTxType(AlloyOpTxType);
+
+impl From<OpTxType> for U8 {
+    fn from(tx_type: OpTxType) -> Self {
+        U8::from(u8::from(tx_type))
+    }
+}
+
+impl TryFrom<u8> for OpTxType {
+    type Error = Eip2718Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        AlloyOpTxType::try_from(value)
+            .map(OpTxType)
+            .map_err(|_| Eip2718Error::UnexpectedType(value))
+    }
+}
+
+impl Default for OpTxType {
+    fn default() -> Self {
+        OpTxType(AlloyOpTxType::Legacy)
+    }
+}
+
+impl PartialEq<u8> for OpTxType {
+    fn eq(&self, other: &u8) -> bool {
+        let self_as_u8: u8 = (*self).into();
+        &self_as_u8 == other
+    }
+}
+
+impl TryFrom<u64> for OpTxType {
+    type Error = Eip2718Error;
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        if value > u8::MAX as u64 {
+            return Err(Eip2718Error::UnexpectedType(0));
+        }
+        Self::try_from(value as u8)
+    }
+}
+
+impl TryFrom<U64> for OpTxType {
+    type Error = Eip2718Error;
+
+    fn try_from(value: U64) -> Result<Self, Self::Error> {
+        let u64_value: u64 = value.try_into().map_err(|_| Eip2718Error::UnexpectedType(0))?;
+        Self::try_from(u64_value)
+    }
+}
+
+impl Encodable for OpTxType {
+    fn length(&self) -> usize {
+        let value: u8 = (*self).into();
+        value.length()
+    }
+
+    fn encode(&self, out: &mut dyn BufMut) {
+        let value: u8 = (*self).into();
+        value.encode(out);
+    }
+}
+
+impl Decodable for OpTxType {
+    fn decode(buf: &mut &[u8]) -> Result<Self, alloy_rlp::Error> {
+        // Decode the u8 value from RLP
+        let value = if buf.is_empty() {
+            return Err(alloy_rlp::Error::InputTooShort);
+        } else if buf[0] == 0x80 {
+            0 // Special case: RLP encoding for integer 0 is `b"\x80"`
+        } else {
+            u8::decode(buf)?
+        };
+
+        OpTxType::try_from(value).map_err(|_| alloy_rlp::Error::Custom("Invalid transaction type"))
+    }
+}
+
+impl TxType for OpTxType {}
```

Comment:
instead of doing this, it makes more sense to add a blanket implementation of `TxType` for all types `T` that impl its super traits. checkout out how it's done for the `FullBlock` trait for example.

Reply:
addressed in f0b692d3f8343e8a2566fb60161ab54483545814